Added a component that adds mesh from a file

joshuahansel opened the pull request:
This does the following:

- Extracted intermediate class `GeneratedMeshComponent` out of `GeometricalComponent`
- Added `FileMeshComponent` as a new child of `GeometricalComponent`
- Extracted interface class `HeatStructureInterface` out of `HeatStructureBase` and inherited from it.
- Inherited `HeatStructureFromFile3D` from `FileMeshComponent` and `HeatStructureInterface`. `HeatStructureBase` is now for generated 2D heat structures only. This requires some checks/casts to change.

This fixes this inheritance:
```
Component -> GeometricalComponent -> Component2D -> HeatStructureBase -> HeatStructureFromFile3D
```
which is obviously wrong (and can be seen to be very wrong from all of the `suppressParameter` calls in its `validParams`). I plan to have an interface class `HeatStructureInterface` instead of the main line base class `HeatStructureBase`.

Now it's
```
Component -> GeometricalComponent -> FileMeshComponent -> HeatStructureFromFile3D
```

Refs #22354

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1265824/) on ac36e5f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22952/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22952/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4b1ed443d5ade609c23d4351cc825026fae6ad61`

joshuahansel committed:
Created secondary base class GeneratedMeshComponent

Components that have a mesh can either generate their own mesh
or load it from a file.

Refs #22354

joshuahansel committed:
Added FileMeshComponent

Refs #22354

joshuahansel committed:
Skipped recover tests for FileMeshComponent

Refs #22354

joshuahansel committed:
Only extracted HS sides for applicable components

Also, removed redundant method.

Refs #22354

joshuahansel committed:
Based HeatStructureFromFile3D on FileMeshComponent

Refs #22354

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1266859/) on b6bfc40 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>0674f2</th>
      <th colspan=4>#22952 b6bfc4</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.51%</td>
      <td>84.51%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>81224</td>
      <td>81222</td>
      <td>-2</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14891</td>
      <td>14893</td>
      <td>+2</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22952/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22952/coverage/framework)


## Modules coverage

### Thermal hydraulics

<table>
  <thead>
    <tr>
      <th></th>
      <th>0674f2</th>
      <th colspan=4>#22952 b6bfc4</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>91.19%</td>
      <td>91.23%</td>
      <td>+0.04%</td>
      <td>98.58%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12399</td>
      <td>12480</td>
      <td>+81</td>
      <td>139</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1198</td>
      <td>1200</td>
      <td>+2</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22952/coverage_diff/thermal_hydraulics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22952/coverage/thermal_hydraulics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22952/coverage/xfem)

</details>



This comment will be updated on new commits.

joshuahansel commented:
This is expected to break Sockeye. Here's the fix: https://hpcgitlab.inl.gov/idaholab/sockeye/-/merge_requests/555.

loganharbour commented:
@joshuahansel category labels go on issues

GiudGiud commented on modules/thermal_hydraulics/src/components/HeatStructureFromFile3D.C:
```diff
@@ -17,221 +16,63 @@ registerMooseObject("ThermalHydraulicsApp", HeatStructureFromFile3D);
+  InputParameters params = FileMeshComponent::validParams();
+  params += HeatStructureInterface::validParams();
+
+  params.addClassDescription("Heat structure component that loads the mesh from an ExodusII file");
```
```suggestion
  params.addClassDescription("Heat structure component that loads a 3D mesh from an ExodusII file");
```

GiudGiud reviewed (commented):
havent looked at all the files yet

I think we may be able to couple already. I ll give it a try

GiudGiud commented on modules/thermal_hydraulics/src/components/FileMeshComponent.C:
```diff
@@ -0,0 +1,199 @@
+  // Generate and set the boundary name
+  for (const auto & id_and_name : new_ids_to_names)
+    boundary_info.sideset_name(id_and_name.first) = genName(_name, id_and_name.second);
+
```
the nodeset names are missing. the nodesets are properly passed though

joshuahansel replied:
I just tried adding this in this loop:
```
boundary_info.nodeset_name(id_and_name.first) = genName(_name, id_and_name.second);
```
but to no effect. The sidesets and nodesets are sharing the IDs, right?

GiudGiud replied:
they should

joshuahansel replied:
Yeah, the IDs in Paraview match. It appears to be not just a problem in this component, but other ones as well. And sometimes it's the other way around - the side sets being named as "surface_#". I'll look into this to see if I can see some kind of pattern...

joshuahansel replied:
Ok, so apparently some gold files have "surface_#", but if I rerun those same tests, I get "nodelist_#" instead. And this appears to happen if we do
```
boundary_info.add_side(elem, side, id);
boundary_info.sideset_name(id) = name;
```
but not when we do
```
boundary_info.add_node(node, id);
boundary_info.nodeset_name(id) = name;
```

I found that this is what fixes it:
```
  for (const auto & id_and_name : new_ids_to_names)
  {
    const std::string sideset_name = genName(_name, id_and_name.second);
    boundary_info.sideset_name(id_and_name.first) = sideset_name;
    boundary_info.nodeset_name(id_and_name.first) = sideset_name;
  }

  boundary_info.build_node_list_from_side_list();
```
It doesn't seem to do anything if I omit the `nodeset_name` line, which surprised me.

`build_node_list_from_side_list()` is already getting executed elsewhere afterwards, so I don't know I need the extra call here, but whatever.

joshuahansel commented on modules/thermal_hydraulics/include/components/GeneratedMeshComponent.h:
```diff
@@ -0,0 +1,52 @@
+                               std::vector<Real> & local_node_locations);
+
+public:
+  static InputParameters validParams();
```
THM has it at the bottom historically, but I've been moving it up in my new classes. Just forgot here - I'll change it.

joshuahansel committed:
Addressed review

Refs #22354

GiudGiud reviewed (approved):
works for heat conduction.

Cant get non-null coupling on INS or INSAD but it's unrelated (missing vector classes for second) reasons

We should work towards FE-FV coupling next

GiudGiud commented:
Adding a valgrind recipe just in case
